gems/aws-sdk-s3/lib/aws-sdk-s3/directory_uploader.rb

```diff
@@ -0,0 +1,215 @@
+  module S3
+    # Raised when DirectoryUploader fails to upload files to S3 bucket
+    class DirectoryUploadError < StandardError
+      def initialize(message, errors = [])
+        @errors = errors
+        super(message)
+      end
+
+      # @return [Array<StandardError>] The list of errors encountered when uploading files
+      attr_reader :errors
+    end
+
+    # @api private
+    class DirectoryUploader
+      def initialize(options = {})
+        @client = options[:client] || Client.new
+        @executor = options[:executor] || DefaultExecutor.new
+        @options = options
+        @abort_upload = false
+      end
+
+      # @return [Client]
+      attr_reader :client
+
+      # @return [Boolean]
+      attr_accessor :abort_upload
+
+      # TODO: Need to add progress tracker
+      def upload(source, bucket:, **options)
+        raise ArgumentError, 'Invalid directory' unless Dir.exist?(source)
+
+        upload_opts = options.dup
+        @ignore_failure = upload_opts.delete(:ignore_failure) || false
+        @errors = []
+
+        uploader = FileUploader.new(
+          multipart_threshold: upload_opts.delete(:multipart_threshold),
+          client: @client,
+          executor: @executor
+        )
+        producer = FileProducer.new(source, build_producer_opts(upload_opts))
+        producer.run
+        uploads = process_upload_queue(producer, uploader, upload_opts.merge(bucket: bucket))
+        build_result(uploads)
+      ensure
+        @executor.shutdown unless @options[:executor]
+      end
+
+      private
+
+      def build_producer_opts(opts)
+        {
+          directory_uploader: self,
+          s3_prefix: opts.delete(:s3_prefix),
+          recursive: opts.delete(:recursive),
+          follow_symlinks: opts.delete(:follow_symlinks),
+          filter_callback: opts.delete(:filter_callback),
+          ignore_failure: @ignore_failure,
+          errors: @errors
+        }
+      end
+
+      def build_result(upload_count)
+        uploads = [upload_count - @errors.count, 0].max
+
+        if @abort_upload
+          msg = "failed to upload directory: uploaded #{uploads} files " \
+            "and failed to upload #{@errors.count} files."
+          raise DirectoryUploadError.new(msg, @errors)
+        else
+          result = { completed_uploads: uploads, failed_uploads: @errors.count }
+          result[:errors] = @errors if @errors.any?
+          result
+        end
+      end
+
+      def process_upload_queue(producer, uploader, opts)
+        upload_attempts = 0
+        completion_queue = Queue.new
+        queue_executor = DefaultExecutor.new
+        while (file = producer.file_queue.shift) != :done
+          break if @abort_upload
+
+          upload_attempts += 1
+          queue_executor.post(file) do |f|
+            uploader.upload(f[:path], opts.merge(key: f[:key]))
+          rescue StandardError => e
+            @errors << e
+            @abort_upload = true unless @ignore_failure
```

Comment:
naming nit - what about `upload_aborted` instead to make it clear that this is state rather than configuration